README.md
# Restivus [v1.0.0](https://github.com/vatfree/meteor-restivus/blob/devel/CHANGELOG.md#change-log)

#### REST APIs for the Best of Us!

**NOTE: This package is based on [nimble:restivus](https://github.com/kahmali/meteor-restivus) created by [Kahmali Rose](https://github.com/kahmali) but has been decaffeinated and expanded.**

Restivus makes building REST APIs in Meteor 1.10.0+ easier than ever before! The package is inspired
by [RestStop2][reststop2-docs] and [Collection API](https://github.com/crazytoad/meteor-collectionapi),
and is built on top of [Simple JSON Routes][json-routes] to provide:

- A simple interface for creating REST APIs
- Easy setup of CRUD endpoints for Mongo Collections
- User authentication via the API
  - Optional login and logout endpoints
  - Access to `this.user` in authenticated endpoints
  - Custom authentication if needed
- Role permissions for limiting access to specific endpoints
  - Works alongside the [`alanning:roles`][alanning-roles] package - Meteor's accepted role
    permission package
- API versioning via URL path
- And also planned:
  - JSON PATCH support on collections
  - Autogenerated OPTIONS endpoint on routes
  - Pre and post hooks on all endpoints

### [Version 1.0.0 Released!](#upgrading-to-100)

## Table of Contents

- [Getting Started](#getting-started)
  - [Installation](#installation)
  - [Quick Start](#quick-start)
  - [Terminology](#terminology)
- [Writing a Restivus API](#writing-a-restivus-api)
  - [Configuration Options](#configuration-options)
  - [Defining Collection Routes](#defining-collection-routes)
    - [Collection](#collection)
    - [Collection Options](#collection-options)
      - [Route Configuration](#route-configuration)
      - [Endpoint Configuration](#endpoint-configuration)
    - [Request and Response Structure](#request-and-response-structure)
    - [Users Collection Endpoints](#users-collection-endpoints)
  - [Defining Custom Routes](#defining-custom-routes)
    - [Path Structure](#path-structure)
    - [Route Options](#route-options)
    - [Defining Endpoints](#defining-endpoints)
      - [Endpoint Configuration](#endpoint-configuration-1)
      - [Endpoint Context](#endpoint-context)
      - [Response Data](#response-data)
  - [Versioning an API](#versioning-an-api)
  - [Documenting an API](#documenting-an-api)
- [Consuming a Restivus API](#consuming-a-restivus-api)
  - [Basic Usage](#basic-usage)
  - [Authenticating](#authenticating)
    - [Default Authentication](#default-authentication)
      - [Logging In](#logging-in)
      - [Logging Out](#logging-out)
      - [Authenticated Calls](#authenticated-calls)
- [Upgrading Restivus](#upgrading-restivus)
  - [Upgrading to 1.10.0](#upgrading-to-100)
  - [Upgrading to 0.8.0](#upgrading-to-080)
  - [Upgrading to 0.7.0](#upgrading-to-070)
  - [Upgrading to 0.6.1](#upgrading-to-061)
- [Resources](#resources)
  - [Plugins](#plugins) 
  - [Change Log](#change-log)
  - [Contributing](#contributing)

# Getting Started

## Installation

You can install Restivus using Meteor's package manager:
```bash
> meteor add vatfree:restivus
```

## Quick Start

Often, the easiest way to explain something is by example, so here's a short example of what it's
like to create an API with Restivus (keep scrolling for a JavaScript version):

###### JavaScript:
```javascript
Items = new Mongo.Collection('items');
Articles = new Mongo.Collection('articles');

if (Meteor.isServer) {
  // Global API configuration
  const Api = new Restivus({
    useDefaultAuth: true,
    prettyJson: true
  });

  // Generates: GET, POST on /api/items and GET, PUT, PATCH, DELETE on
  // /api/items/:id for the Items collection
  Api.addCollection(Items);

  // Generates: POST on /api/users and GET, DELETE /api/users/:id for
  // Meteor.users collection
  Api.addCollection(Meteor.users, {
    excludedEndpoints: ['getAll', 'put'],
    routeOptions: {
      authRequired: true
    },
    endpoints: {
      post: {
        authRequired: false
      },
      delete: {
        roleRequired: 'admin'
      }
    }
  });

  // Maps to: /api/articles/:id
  Api.addRoute('articles/:id', {authRequired: true}, {
    get: function () {
      return Articles.findOne(this.urlParams.id);
    },
    delete: {
      roleRequired: ['author', 'admin'],
      action: function () {
        if (Articles.remove(this.urlParams.id)) {
          return {status: 'success', data: {message: 'Article removed'}};
        }
        return {
          statusCode: 404,
          body: {status: 'fail', message: 'Article not found'}
        };
      }
    }
  });
}
```

## Terminology

Just to clarify some terminology that will be used throughout these docs:

_**(HTTP) Method:**_
- The type of HTTP request (e.g., GET, PUT, POST, etc.)

_**Endpoint:**_
- The function executed when a request is made at a given URL path for a specific HTTP method

_**Route:**_
- A URL path and its set of configurable endpoints

# Writing A Restivus API

**Restivus is a _server-only_ package. Attempting to access any of its methods from the client will
result in an error.**

## Configuration Options

The following configuration options are available when initializing an API using
`new Restivus(options)`:

##### `apiPath`
- _String_
- Default: `'api/'`
- The base path for your API. If you use `'api'` and add a route called `'users'`, the URL will be
  `https://yoursite.com/api/users/`.

##### `auth`
- _Object_
  - `token`  _String_
    - Default: `'services.resume.loginTokens.hashedToken'`
    - The path to the hashed auth token in the `Meteor.user` document. This location will be checked
      for a matching token if one is returned in `auth.user()`.
  - `user`  _Function_
    - Default: Get user ID and auth token from `X-User-Id` and `X-Auth-Token` headers
        ```javascript
        function() {
          return {
            userId: this.request.headers['x-user-id'],
            token: Accounts._hashLoginToken(this.request.headers['x-auth-token'])
          };
        }
        ```

    - Provides one of two levels of authentication, depending on the data returned. The context
      within this function is the [endpoint context](#endpoint-context) without `this.user` and
      `this.userId` (well, that's what we're working on here!). Once the user authentication
      completes successfully, the authenticated user and their ID will be attached to the [endpoint
      context](#endpoint-context). The two levels of custom authentication and their required return
      data are:
      - Partial auth
        - `userId`: The ID of the user being authenticated
        - `token`: The auth token to be verified
        - If both a `userId` and `token` are returned, Restivus will hash the token, then,
          authentication will succeed if the `hashedToken`
          exists in the given `Meteor.user` document at the location specified in `auth.token`
      - Complete auth
        - `user`: The fully authenticated `Meteor.user`
        - This is your chance to completely override the user authentication process. If a `user` is
          returned, any `userId` and `token` will be ignored, as it's assumed that you have already
          successfully authenticated the user (by whatever means you deem necessary). The given user
          is simply attached to the [endpoint context](#endpoint-context), no questions asked.
          
      For either level of auth described above, you can optionally return a custom error response by 
      providing that response in an `error` field of your response object. The `error` value can be 
      [any valid response](#response-data). If an `error` field exists in the object returned from 
      your custom auth function, all other fields will be ignored. Do **not** provide an `error` 
      value if you intend for the authentication to pass successfully.

##### `defaultHeaders`
- _Object_
- Default: `{ 'Content-Type': 'application/json' }`
- The response headers that will be returned from every endpoint by default. These can be overridden
  by [returning `headers` of the same name from any endpoint](#response-data).

##### `defaultOptionsEndpoint`
- [_Endpoint_](#endpoint-configuration)
- Default: undefined
- If an endpoint is provided, it will be used as the OPTIONS endpoint on all routes, except those
  that have one manually defined. This can be used to DRY up your API, since OPTIONS endpoints will
  frequently [respond generically](http://zacstewart.com/2012/04/14/http-options-method.html) across
  all routes.

##### `enableCors`
- _Boolean_
- Default: `true`
- If true, enables cross-origin resource sharing ([CORS]). This allows your API to receive requests
  from _any_ domain (when `false`, the API will only accept requests from the domain where the API
  is being hosted. _Note: Only applies to requests originating from browsers)._

##### `onLoggedIn`
- _Function_
- Default: `undefined`
- A hook that runs once a user has been successfully logged into their account via the `/login`
  endpoint. [Context](#endpoint-context) is the same as within authenticated endpoints. Any
  returned data will be added to the response body as `data.extra`.

##### `onLoggedOut`
- _Function_
- Default: `undefined`
- Same as onLoggedIn, but runs once a user has been successfully logged out of their account via
  the `/logout` endpoint. [Context](#endpoint-context) is the same as within authenticated
  endpoints. Any returned data will be added to the response body as `data.extra`.

##### `onAuth`
- _Function_
- Default: `undefined`
- Function called right after authorization has been checked and before the route `action()` is called. This
  can be used to add extra checks to the authorization. The new 

##### `onAction`
- _Function_
- Default: `undefined`
- Function called right before the `action()` on the route is called. This allows to capture all calls to the api
  and potentially manipulate the endpoint context that is passed to the route.

##### `onReturn`
- _Function_
- Default: `undefined`
- Function called right before the result is returned to the client. This allows to capture the return value

CHANGELOG.md
# Change Log

## [v1.1.0] - 2020-12-17

#### Added
onAuth, onAction, onReturn functions to global options.

## [v1.0.0] - 2020-12-10

#### Changed
Decaffeinated project, added ecmascript support, expanded usage of `roleRequired` to allow setting of scope (or group).

## [v0.8.12] - 2017-01-23

#### Added
- Support for SHA-256 hashed passwords sent from client in the [default login endpoint](https://github.com/kahmali/meteor-restivus#logging-in)
- [PATCH support](https://github.com/kahmali/meteor-restivus#patch) for default collection endpoints 

#### Changed
- Allow custom error responses to be returned from [custom auth functions](https://github.com/kahmali/meteor-restivus#auth) by providing an `error` value in return object

#### Fixed 
- Bcrypt build error (required upgrade of `accounts-password` dependency to v1.3.3)


## [v0.8.11] - 2016-06-16

#### Fixed
- Allow null or undefined to be returned from endpoint function when HTTP response is handled manually ([#135][])


## [v0.8.10] - 2016-03-28

#### Changed
- Add result from `onLoggedOut` hook to `data.extra` field ([#196][]) 
- Remove the default `onLoggedIn` and `onLoggedOut` hooks that returned an empty object (docs incorrectly listed them as `undefined`, but the result was not being used prior to [v0.8.9]) 


## [v0.8.9] - 2016-03-14

#### Changed
- Add result from `onLoggedIn` hook to `data.extra` field ([#194][]) 


## [v0.8.8] - 2016-03-10

#### Changed
- Update to the latest version of `simple:json-routes` ([v2.1.0](https://github.com/stubailo/meteor-rest/tree/master/packages/json-routes#210))
  - Resolves errors when using alongside Iron Router ([#185][]) 


## [v0.8.7] - 2016-02-24

#### Fixed
- Declare dependency on accounts-password ([#169][])


## [v0.8.6] - 2016-02-24

#### Changed
- Update to the latest version of `simple:json-routes` ([v2.0.1](https://github.com/stubailo/meteor-rest/tree/master/packages/json-routes#201))
  - Increases max request size to 50mb ([#78][] and [#179][]) 


## [v0.8.5] - 2016-02-24

#### Fixed
- Remove `Route` class from global namespace ([#170][])


## [v0.8.4] - 2015-08-18

#### Added
- Default OPTIONS endpoint when CORS is enabled ([#99][])

#### Fixed
- Enable CORS with default authentication when both are configured ([#99][])


## [v0.8.3] - 2015-08-12

#### Changed
- Accept explicit `username` or `email` in default login endpoint ([#121][])


## [v0.8.2] - 2015-08-05

#### Changed
- Update `simple:json-routes` to v1.0.4 to remove Connect deprecation warnings ([#118][])


## [v0.8.1] - 2015-07-07

#### Added 
- Default logout endpoint at `POST /api/logout`

#### Deprecated
- Default logout endpoint at `GET /api/logout` (#100)
  - Will be removed in v1.0
  - Use `POST /api/logout` instead


## [v0.8.0] - 2015-07-06

**_WARNING!_ API-breaking changes! Please be aware when upgrading!**

_Note: API-breaking changes are in **bold**_

#### Added
- Support [API versioning][versioning an api] via URL path
- API configuration option for defining a default OPTIONS endpoint (used on 
  any route that doesn't define it's own OPTIONS endpoint): `defaultOptionsEndpoint`
    
#### Changed
- **Export Restivus as a "class" instead of an object**
- Replace Iron Router with [JSON Routes]
  (https://github.com/stubailo/meteor-rest/tree/master/packages/json-routes)! ([#38][])
- Return standard response codes in any autogenerated endpoints ([#48][])
  - `201`: Resource created (in `POST` collection endpoints)
  - `403`: Role permission errors
  - `405`: API method not found (but route exists)
- Remove dependency on `useAuth` API config option for endpoint authentication ([#49][])
  
#### Deprecated
- `useAuth` API config option (replaced with `useDefaultAuth` to better reflect it's purpose, which 
  is to generate the default auth endpoints)

#### Fixed
- Iron Router related issues
  - Issue [#24][]: Interference with other routers
  - Issue [#35][]: App hangs on iOS
  - Issue [#43][]: Deployed app shows `iron:router` welcome screen
- Issue [#91][]: Return a `401` error when auth info is not provided in endpoint requiring default 
  auth (previously returned `500` error) 
  
#### Removed
- **`deleteAll` collection endpoint (it had the potential to be quite destructive - [#47][])**
- **`Restivus.configure()` (configuration now available via Restivus constructor)**


## [v0.7.1] - 2015-07-01

#### Fixed
- Issue #95: On 401 and 403 errors, Restivus should wait before sending any data


## [v0.7.0] - 2015-06-18

**_WARNING!_ Potentially breaking changes! Please be aware when upgrading!**

#### Changed
- Update default auth endpoints to match current Accounts token storage ([#79][])
  - **_WARNING!_ All clients consuming a Restivus API _with the default authentication_ will need to 
    reauthenticate after this update**
  - Login token is now stored as `hashedToken` instead of `token`
- Return `401 Unauthorized` for failed authentication
- When logging in with bad credentials, randomly delay the response ([#81][])
- Declare dependency on `accounts-base` package
  

## [v0.6.6] - 2015-05-25
  

lib/restivus.js
import { _ } from 'underscore';
import { Accounts } from 'meteor/accounts-base';
import { Route } from './route';
import { Auth } from './auth';

export const Restivus = function (options) {
    let corsHeaders;
    this._routes = [];
    this._config = {
        paths: [],
        useDefaultAuth: false,
        apiPath: 'api/',
        version: null,
        prettyJson: false,
        auth: {
            token: 'services.resume.loginTokens.hashedToken',
            user: function () {
                var token;
                if (this.request.headers['x-auth-token']) {
                    token = Accounts._hashLoginToken(this.request.headers['x-auth-token']);
                }
                return {
                    userId: this.request.headers['x-user-id'],
                    token: token
                };
            }
        },
        defaultHeaders: {
            'Content-Type': 'application/json'
        },
        enableCors: true
    };
    _.extend(this._config, options);

    if (this._config.enableCors) {
        corsHeaders = {
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Headers': 'Origin, X-Requested-With, Content-Type, Accept'
        };

        if (this._config.useDefaultAuth) {
            corsHeaders['Access-Control-Allow-Headers'] += ', X-User-Id, X-Auth-Token';
        }
        _.extend(this._config.defaultHeaders, corsHeaders);

        if (!this._config.defaultOptionsEndpoint) {
            this._config.defaultOptionsEndpoint = function () {
                this.response.writeHead(200, corsHeaders);
                return this.done();
            };
        }
    }

    if (this._config.apiPath[0] === '/') {
        this._config.apiPath = this._config.apiPath.slice(1);
    }

    if (_.last(this._config.apiPath) !== '/') {
        this._config.apiPath = this._config.apiPath + '/';
    }

    if (this._config.version) {
        this._config.apiPath += this._config.version + '/';
    }

    if (this._config.useDefaultAuth) {
        this._initAuth();
    } else if (this._config.useAuth) {
        this._initAuth();
        console.warn('Warning: useAuth API config option will be removed in Restivus v1.0 ' + '\n    Use the useDefaultAuth option instead');
    }

    return this;
};

/**
 Add endpoints for the given HTTP methods at the given path

 @param path {String} The extended URL path (will be appended to base path of the API)
 @param options {Object} Route configuration options
 @param options.authRequired {Boolean} The default auth requirement for each endpoint on the route
 @param options.roleRequired {String or String[]} The default role required for each endpoint on the route
 @param endpoints {Object} A set of endpoints available on the new route (get, post, put, patch, delete, options)
 @param endpoints.<method> {Function or Object} If a function is provided, all default route
 configuration options will be applied to the endpoint. Otherwise an object with an `action`
 and all other route config options available. An `action` must be provided with the object.
 */
Restivus.prototype.addRoute = function (path, options, endpoints) {
    const route = new Route(this, path, options, endpoints);
    this._routes.push(route);
    route.addToApi();

    return this;
};

/**
 Generate routes for the Meteor Collection with the given name
 */
Restivus.prototype.addCollection = function (collection, options) {
    if (options == null) {
        options = {};
    }

    const methods = ['get', 'post', 'put', 'patch', 'delete', 'getAll'];
    const methodsOnCollection = ['post', 'getAll'];

    let collectionEndpoints,
        collectionRouteEndpoints,
        endpointsAwaitingConfiguration,
        entityRouteEndpoints,
        excludedEndpoints;

    if (collection === Meteor.users) {
        collectionEndpoints = this._userCollectionEndpoints;
    } else {
        collectionEndpoints = this._collectionEndpoints;
    }
    endpointsAwaitingConfiguration = options.endpoints || {};

    const routeOptions = options.routeOptions || {};
    excludedEndpoints = options.excludedEndpoints || [];

    const path = options.path || collection._name;
    collectionRouteEndpoints = {};
    entityRouteEndpoints = {};

    if (_.isEmpty(endpointsAwaitingConfiguration) && _.isEmpty(excludedEndpoints)) {
        _.each(methods, function (method) {
            if (methodsOnCollection.indexOf(method) >= 0) {
                _.extend(collectionRouteEndpoints, collectionEndpoints[method].call(this, collection));
            } else {
                _.extend(entityRouteEndpoints, collectionEndpoints[method].call(this, collection));
            }
        }, this);
    } else {
        _.each(methods, function (method) {
            if (excludedEndpoints.indexOf(method) < 0 && endpointsAwaitingConfiguration[method] !== false) {
                const endpointOptions = endpointsAwaitingConfiguration[method];
                const configuredEndpoint = {};
                _.each(collectionEndpoints[method].call(this, collection), function (action, methodType) {
                    return configuredEndpoint[methodType] = _.chain(action).clone().extend(endpointOptions).value();
                });
                if (methodsOnCollection.indexOf(method) >= 0) {
                    _.extend(collectionRouteEndpoints, configuredEndpoint);
                } else {
                    _.extend(entityRouteEndpoints, configuredEndpoint);
                }
            }
        }, this);
    }

    this.addRoute(path, routeOptions, collectionRouteEndpoints);
    this.addRoute(path + '/:id', routeOptions, entityRouteEndpoints);

    return this;
};

/**
 * Creates a MongoDB selector query from the request queries.
 * @param queryParams queryParams object from the request object
 * @returns selector object for the MongoDB
 */
const getQueryParameters = function (queryParams) {
    const selector = {};
    Object.keys(queryParams).forEach(function (key, index) {
        if (isNaN(Number(queryParams[key]))) {
            selector[key] = queryParams[key];
        } else {
            selector[key] = Number(queryParams[key]);
        }
    });

    return selector;
};

/**
 A set of endpoints that can be applied to a Collection Route
 */
Restivus.prototype._collectionEndpoints = {
    get: function (collection) {
        return {
            get: {
                action: function () {
                    const selector = getQueryParameters(this.queryParams);
                    selector._id = this.urlParams.id;

                    const entity = collection.find(selector).fetch();
                    if (!_.isEmpty(entity)) {
                        return {
                            status: 'success',
                            data: entity
                        };
                    } else {
                        return {
                            statusCode: 204,
                            body: {
                                status: 'fail',
                                message: 'Item not found'
                            }
                        };
                    }
                }
            }
        };
    },
    put: function (collection) {
        return {
            put: {
                action: function () {
                    const selector = getQueryParameters(this.queryParams);
                    selector._id = this.urlParams.id;

                    const entityIsUpdated = collection.update(selector, this.bodyParams);
                    if (entityIsUpdated) {
                        const entity = collection.findOne(this.urlParams.id);
                        return {
                            status: 'success',
                            data: entity
                        };
                    } else {
                        return {
                            statusCode: 400,
                            body: {
                                status: 'fail',
                                message: 'Item not updated'
                            }
                        };
                    }
                }
            }
        };
    },
    patch: function (collection) {
        return {
            patch: {
                action: function () {
                    const selector = getQueryParameters(this.queryParams);
                    selector._id = this.urlParams.id;

                    const entityIsUpdated = collection.update(selector, {
                        $set: this.bodyParams
                    });
                    if (entityIsUpdated) {
                        const entity = collection.findOne(selector);
                        return {
                            status: 'success',
                            data: entity
                        };
                    } else {
                        return {
                            statusCode: 400,
                            body: {
                                status: 'fail',
                                message: 'Item not updated'
                            }
                        };
                    }
                }
            }
        };
    },
    'delete': function (collection) {
        return {
            'delete': {
                action: function () {
                    const selector = getQueryParameters(this.queryParams);
                    selector._id = this.urlParams.id;

                    if (collection.remove(selector)) {
                        return {
                            status: 'success',
                            data: {
                                message: 'Item removed'
                            }
                        };
                    } else {
                        return {
                            statusCode: 400,
                            body: {
                                status: 'fail',
                                message: 'Could not delete item'
                            }
                        };
                    }
                }
            }
        };
    },
    post: function (collection) {
        return {
            post: {
                action: function () {
                    const entityId = collection.insert(this.bodyParams);
                    const entity = collection.findOne(entityId);
                    if (!_.isEmpty(entity)) {
                        return {
                            statusCode: 201,
                            body: {
                                status: 'success',
                                data: entity
                            }
                        };
                    } else {
                        return {
                            statusCode: 400,
                            body: {
                                status: 'fail',
                                message: 'No item added'
                            }
                        };
                    }
                }
            }
        };
    },
    getAll: function (collection) {
        return {
            get: {
                action: function () {
                    const selector = getQueryParameters(this.queryParams);
                    const entities = collection.find(selector).fetch();
                    if (entities) {
                        return {
                            status: 'success',
                            data: entities
                        };
                    } else {
                        return {
                            statusCode: 204,
                            body: {
                                status: 'fail',
                                message: 'Unable to retrieve items from collection'
                            }
                        };
                    }
                }
            }
        };
    }
};

/**
 A set of endpoints that can be applied to a Meteor.users Collection Route
 */
Restivus.prototype._userCollectionEndpoints = {
    get: function (collection) {
        return {
            get: {
                action: function () {
                    const entity = collection.findOne(this.urlParams.id, {
                        fields: {
                            profile: 1
                        }
                    });
                    if (entity) {
                        return {
                            status: 'success',
                            data: entity
                        };
                    } else {
                        return {
                            statusCode: 404,
                            body: {
                                status: 'fail',
                                message: 'User not found'
                            }
                        };
                    }
                }
            }
        };
    },
    put: function (collection) {
        return {
            put: {
                action: function () {
                    const entityIsUpdated = collection.update(this.urlParams.id, {
                        $set: {
                            profile: this.bodyParams
                        }
                    });
                    if (entityIsUpdated) {
                        const entity = collection.findOne(this.urlParams.id, {
                            fields: {
                                profile: 1
                            }
                        });
                        return {
                            status: 'success',
                            data: entity
                        };
                    } else {
                        return {
                            statusCode: 404,
                            body: {
                                status: 'fail',
                                message: 'User not found'
                            }
                        };
                    }
                }
            }
        };
    },
    'delete': function (collection) {
        return {
            'delete': {
                action: function () {
                    if (collection.remove(this.urlParams.id)) {
                        return {
                            status: 'success',
                            data: {
                                message: 'User removed'
                            }
                        };
                    } else {
                        return {
                            statusCode: 404,
                            body: {
                                status: 'fail',
                                message: 'User not found'
                            }
                        };
                    }
                }
            }
        };
    },
    post: function (collection) {
        return {
            post: {
                action: function () {
                    const entityId = Accounts.createUser(this.bodyParams);
                    const entity = collection.findOne(entityId, {
                        fields: {
                            profile: 1
                        }
                    });
                    if (entity) {
                        return {
                            statusCode: 201,
                            body: {
                                status: 'success',
                                data: entity
                            }
                        };
                    } else {
                        return ({
                            statusCode: 400,
                            body: {
                                status: 'fail',
                                message: 'No user added'
                            }
                        });
                    }
                }
            }
        };
    },
    getAll: function (collection) {
        return {
            get: {
                action: function () {
                    const entities = collection.find({}, {
                        fields: {
                            profile: 1
                        }
                    }).fetch();
                    if (entities) {
                        return {
                            status: 'success',
                            data: entities
                        };
                    } else {
                        return {
                            statusCode: 404,
                            body: {
                                status: 'fail',
                                message: 'Unable to retrieve users'
                            }
                        };
                    }
                }
            }
        };
    }
};

/*
 Add /login and /logout endpoints to the API
 */
Restivus.prototype._initAuth = function () {
    const self = this;

    /*
     Add a login endpoint to the API

     After the user is logged in, the onLoggedIn hook is called (see Restfully.configure() for
     adding hook).
     */
    this.addRoute('login', {
        authRequired: false
    }, {
        post: function () {
            let auth;

            const user = {};
            if (this.bodyParams.user) {
                if (this.bodyParams.user.indexOf('@') === -1) {
                    user.username = this.bodyParams.user;
                } else {
                    user.email = this.bodyParams.user;
                }
            } else if (this.bodyParams.username) {
                user.username = this.bodyParams.username;
            } else if (this.bodyParams.email) {
                user.email = this.bodyParams.email;
            }

            let password = this.bodyParams.password;
            if (this.bodyParams.hashed) {
                password = {
                    digest: password,
                    algorithm: 'sha-256'
                };
            }

            try {
                auth = Auth.loginWithPassword(user, password);
            } catch (e) {
                return {
                    statusCode: e.error,
                    body: {
                        status: 'error',
                        message: e.reason
                    }
                };
            }

            if (auth.userId && auth.authToken) {
                const searchQuery = {};
                searchQuery[self._config.auth.token] = Accounts._hashLoginToken(auth.authToken);
                this.user = Meteor.users.findOne({
                    '_id': auth.userId
                }, searchQuery);

                this.userId = this.user ? this.user._id : null;
            }

            const response = {
                status: 'success',
                data: auth
            };

            const onLoggedIn = self._config.onLoggedIn;
            const extraData = _.isFunction(onLoggedIn) ? onLoggedIn.call(this) : null;
            if (extraData != null) {
                _.extend(response.data, {
                    extra: extraData
                });
            }

            return response;
        }
    });

    const logout = function () {
        const authToken         = this.request.headers['x-auth-token'];
        const hashedToken       = Accounts._hashLoginToken(authToken);
        const tokenLocation     = self._config.auth.token;
        const index             = tokenLocation.lastIndexOf('.');
        const tokenPath         = tokenLocation.substring(0, index);
        const tokenFieldName    = tokenLocation.substring(index + 1);
        const tokenToRemove     = {};
        tokenToRemove[tokenFieldName] = hashedToken;

        const tokenRemovalQuery = {};
        tokenRemovalQuery[tokenPath] = tokenToRemove;

        Meteor.users.update(this.user._id, {
            $pull: tokenRemovalQuery
        });

        const response = {
            status: 'success',
            data: {
                message: 'You\'ve been logged out!'
            }
        };
        const onLoggedOut = self._config.onLoggedOut;
        const extraData = _.isFunction(onLoggedOut) ? onLoggedOut.call(this) : null;
        if (extraData != null) {
            _.extend(response.data, {
                extra: extraData
            });
        }

        return response;
    };

    /*
     Add a logout endpoint to the API

     After the user is logged out, the onLoggedOut hook is called (see Restfully.configure() for
     adding hook).
     */
    return this.addRoute('logout', {
        authRequired: true
    }, {
        get: function () {
            console.warn('Warning: Default logout via GET will be removed in Restivus v1.0. Use POST instead.');
            console.warn('    See https://github.com/kahmali/meteor-restivus/issues/100');
            return logout.call(this);
        },
        post: logout
    });
};
